### DISCUSSION TITLE
Demo commit-linter Args appear empty even with commit message argument

### DISCUSSION THREAD

**Discussion by Rvanmech** [alirezanet/Husky.Net#64] [Q&A] [answered]
Hi there,

We have been using this Husky.Net in task-runner.json to run things like dotnet format, dotnet build and dotnet test, so thank you for creating it. Lately I have been experimenting trying to add commit message linting, so I tried using the excellent example posted. 

However, I'm running into a problem where using example provided is not succeeding fully. I have Husky 0.5.4 installed, and when I go to do the commit to kick off the `.husky/csx/commit-lint.csx` file, it ends up showing the error `Unable to execute cached script.`

As far as I can tell, the error is being thrown when attempting to do `File.ReadAllLines(Arg[0])[0]`, and it seems Args[0] has a Count of 1, but must be either a whitespace character or empty string since it prints out nothing. 

https://github.com/alirezanet/Husky.Net/blob/master/.husky/csx/commit-lint.csx#L10

I really don't think I have changed things, since my task-runner.json has the following, but it really seems like the commit message is always coming back as empty. 

``` json
{
      "name": "commit-message-linter",
      "command": "dotnet",
      "group": "pre-commit",
      "args": ["husky", "exec", ".husky/csx/commit-lint.csx", "--args", "${args}"]
    },
```
For the record I invoke the script by `git commit -m "Change message"` and it seems to not be picking up the message contents. 
If I invoke it directly like 
`dotnet husky run --name "commit-message-linter" --args "test"` from the terminal, then the text "test" does show up. 
So I my next thought was to look into`${args}` in task-runner.json to ensure it is capturing the commit message string and passing it along correctly. Does anyone have an idea of what might be going wrong?
Thank you.

Other relevant info: I'm on a macOS Monterey, using zsh, with .NET6 SDK and runtime installed

**Comment by alirezanet** [comment, marked as answer]
Hi @Rvanmech,
Thank you for the feedback, you are the second person that recently mentioned having problems with macOS, unfortunately, at the moment I don't have access to my mac machine so I can not debug exactly what is wrong. But if I give you some context, I hope you can solve it. 
when you commit something, Git will create a file containing your commit message and it won't pass the commit message directly to the CSX script so the first thing I would check is your commit message file path in the `commit-msg` hook, `dotnet husky run --name "commit-message-linter" --args "$1"`. here the `$1` is not the commit message, it is a file path you must pass to the linter. so make sure you have it. (maybe it is better if you also log the `$1` on your mac `echo "\$1 is $1"` to make sure it contains the file path or not)

also to test your csx file I recommend using the exec command and passing a text file to it to make sure it is working as expected.
`dotnet husky exec  "CSXFILE" --args "TEXT FILE PATH CONTAINING A SAMPLE COMMIT MSG"` 

Let me know if you need further information

**Reply by Rvanmech** [reply]
@alirezanet  Thank you! 
I confirmed that  $1 is blank, i.e. `echo "\$1 is $1"` just prints out nothing "$1 is " 

However, the CSXFile is working when I provide the Text file directly as you suggested as an argument. Doing some more research, I found this simple sh script that I could run on stackoverflow (https://askubuntu.com/a/1428604) and was able to confirm that I can capture the $1 argument by invoking that script. So, that said I tried looking into it, but not sure why that $1 wouldn't be passed along. Do you have any ideas for how to debug further, maybe for whatever calls the pre-commit script?  
One thing I thought was curious is that pre-commit is not a `.sh` script, but trying to rename it to .sh just skipped the pre-commit husky checks and committed directly, so I think that wasn't quite right. 

Thank you.

**Reply by Rvanmech** [reply]
@alirezanet One thing to add: I found the `.husky` script within the `_` directory. I tried echoing `$@` and it appears to be blank too, which that would seem to lead to our situation. Same thing for `$2` and `$3`, so IMO whatever is invoking the .husky script is not appearing to capture those command line args to pass it on. Edit: `$1` too appears empty within the `.husky` script.

**Reply by Rvanmech** [reply]
``` 
% git cm "Trial commit"
husky (debug) - starting pre-commit...
$@ is 
$2 is 
$3 is 
.husky/pre-commit
.husky/_/husky.sh
$1 is 
[Husky] 🚀 Loading tasks ...
--------------------------------------------------
[Husky] ⚡ Preparing task 'commit-message-linter'
[Husky] ⌛ Executing task 'commit-message-linter' ...
Args.Count: 1

Unable to execute cached script.

  ❌ Task 'commit-message-linter' failed in 1,339ms

husky - pre-commit hook exited with code 1 (error)
```

**Reply by alirezanet** [reply]
Hi @Rvanmech,  good information, thanks!
so the problem is related to the hook itself and git which is invoking it in the background. soon I'll do a quick research on why git hooks are behaving differently in macOS and let you know if I found anything.

**Reply by alirezanet** [reply]
until we find a possible solution to fix this, can you change the shell from `#!/bin/sh` to `#!/usr/bin/env sh` or `#!/usr/bin/env bash`  and see if it works? 

--- 
**EDIT -- or more precisely this:**
commit-msg: 
```
#!/usr/bin/env sh
. "$(dirname -- "$0")/_/husky.sh"
```

**Reply by alirezanet** [reply]
> One thing I thought was curious is that pre-commit is not a .sh script, but trying to rename it to .sh just skipped the pre-commit

Yes, the name of the[ git hook](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) is important., git can't pick it up if you change it.

**Reply by Rvanmech** [reply]
@alirezanet  Thanks for the info and help so far! I tried both of the examples, and they are giving me the same result where `$@` and the variables `$1`, `$2` etc are blank. 
I'll go do some research as well to see if I can find anything.

**Reply by Rvanmech** [reply]
@alirezanet  I realized something based off the info you gave me! I have made a user error.
I was inadvertently using the pre-commit hook instead of the commit-msg hook, since I was modifying some code we had already for pre-commit.
So.. the long story is that it works!  

``` 
% git cm "Hello"             
.husky/commit-msg
.husky/_/husky.sh
$1 is .git/COMMIT_EDITMSG
[Husky] 🚀 Loading tasks ...
--------------------------------------------------
[Husky] ⚡ Preparing task 'commit-message-linter'
[Husky] ⌛ Executing task 'commit-message-linter' ...
1
.git/COMMIT_EDITMSG
Invalid commit message
script execution failed
```

**Reply by alirezanet** [reply]
I'm glad you've solved the problem 😊.

**Reply by WeihanLi** [reply]
Oh I met the same problem, I added it to the pre-commit hook, it works when I move it to the commit-msg hook.
Thanks very much
